src/components/table.js

```diff
@@ -0,0 +1,495 @@
+    else
+      return sortUp ? bIndex - aIndex : aIndex - bIndex;
+  };
+
+  // toggles checkbox on/off
+  toggleChecked = (datum, field) => {
+    const newData = copyObject(this.props.data);
+
+    for (const row of newData) {
+      if (compareObjects(row, datum))
+        row[field] = !row[field];
+    }
+
+    this.state.onChange(newData);
+  };
+
+  // solo checkbox (turn all others off)
+  soloChecked = (datum, field) => {
+    const newData = copyObject(this.props.data);
+
+    let allOthersUnchecked = true;
+    for (const row of newData) {
+      if (!compareObjects(row, datum) && row[field]) {
+        allOthersUnchecked = false;
+        break;
+      }
+    }
+
+    for (const row of newData) {
+      if (allOthersUnchecked || compareObjects(row, datum))
+        row[field] = true;
+      else
+        row[field] = false;
+    }
+
+    this.state.onChange(newData);
+  };
+
+  // checks whether all checkboxes are checked
+  allChecked = (field) => {
+    for (const datum of this.props.data) {
+      if (!datum[field])
+        return false;
+    }
+
+    return true;
+  };
+
+  // check or uncheck all checkboxes
+  toggleAll = (field) => {
+    const newData = copyObject(this.props.data);
+
+    const newChecked = !this.allChecked(field);
+    for (const datum of newData)
+      datum[field] = newChecked;
+
+    this.state.onChange(newData);
+  };
+
+  // display component
+  render() {
+    return (
+      <TableContext.Provider
+        value={{
+          data: this.state.data || [],
+          sortField: this.state.sortField,
+          sortUp: this.state.sortUp,
+          toggleChecked: this.toggleChecked,
+          soloChecked: this.soloChecked,
+          allChecked: this.allChecked,
+          toggleAll: this.toggleAll,
+          changeSort: this.changeSort,
+          superContents: this.props.superContents || [],
+          superStyles: this.props.superStyles || [],
+          superClasses: this.props.superClasses || [],
+          superColspans: this.props.superColspans || [],
+          headContents: this.props.headContents || [],
+          headFields: this.props.headFields || [],
+          headStyles: this.props.headStyles || [],
+          headClasses: this.props.headClasses || [],
+          headTooltips: this.props.headTooltips || [],
+          bodyValues: this.props.bodyValues || [],
+          bodyFullValues: this.props.bodyFullValues || [],
+          bodyStyles: this.props.bodyStyles || [],
+          bodyClasses: this.props.bodyClasses || [],
+          bodyTooltips: this.props.bodyTooltips || []
+        }}
+      >
+        <table className={this.props.className}>
+          <thead>
+            <Super />
+            <Head />
+          </thead>
+          <tbody>
+            <Body />
+          </tbody>
+        </table>
+      </TableContext.Provider>
+    );
+  }
+}
+const TableContext = React.createContext({});
+
+// super section
+// row above head row
+class Super extends Component {
```

Comment:
I wonder whether we should use another class name. When I first saw it, I thought it would be a superclass of `supercell`; Plus, `super()` is a built-in function in JS class definition.

Reply:
Yeah I don't know what else to call it. `Top`?